src/core/net/dnssd_server.cpp

```diff
@@ -668,34 +749,386 @@ const Srp::Server::Service *Server::GetNextSrpService(const Srp::Server::Host &
+        break;
+    default:
+        break;
+    }
+
+    return canAnswer;
+}
+
+bool Server::CanAnswerQuery(const Server::QueryTransaction &aQuery, const char *aHostFullName)
+{
+    char         name[Name::kMaxNameSize];
+    DnsQueryType sdType;
+
+    sdType = GetQueryType(aQuery.GetResponseHeader(), aQuery.GetResponseMessage(), name);
+    return (sdType == kDnsQueryResolveHost) && (strcmp(name, aHostFullName) == 0);
+}
+
+void Server::AnswerQuery(QueryTransaction &                aQuery,
+                         const char *                      aServiceFullName,
+                         const otDnssdServiceInstanceInfo &aInstanceInfo)
+{
+    Header &          responseHeader  = aQuery.GetResponseHeader();
+    Message &         responseMessage = aQuery.GetResponseMessage();
+    Error             error           = kErrorNone;
+    NameCompressInfo &compressInfo    = aQuery.GetNameCompressInfo();
+
+    if (HasQuestion(aQuery.GetResponseHeader(), aQuery.GetResponseMessage(), aServiceFullName,
+                    ResourceRecord::kTypePtr))
+    {
+        SuccessOrExit(error = AppendPtrRecord(responseMessage, aServiceFullName, aInstanceInfo.mFullName,
+                                              aInstanceInfo.mTtl, compressInfo));
+        IncResourceRecordCount(responseHeader, false);
+    }
+
+    for (uint8_t additional = 0; additional <= 1; additional++)
+    {
+        if (HasQuestion(aQuery.GetResponseHeader(), aQuery.GetResponseMessage(), aInstanceInfo.mFullName,
+                        ResourceRecord::kTypeSrv) == !additional)
+        {
+            SuccessOrExit(error = AppendSrvRecord(responseMessage, aInstanceInfo.mFullName, aInstanceInfo.mHostName,
+                                                  aInstanceInfo.mTtl, aInstanceInfo.mPriority, aInstanceInfo.mWeight,
+                                                  aInstanceInfo.mPort, compressInfo));
+            IncResourceRecordCount(responseHeader, additional);
+        }
+
+        if (HasQuestion(aQuery.GetResponseHeader(), aQuery.GetResponseMessage(), aInstanceInfo.mFullName,
+                        ResourceRecord::kTypeTxt) == !additional)
+        {
+            SuccessOrExit(error = AppendTxtRecord(responseMessage, aInstanceInfo.mFullName, aInstanceInfo.mTxtData,
+                                                  aInstanceInfo.mTxtLength, aInstanceInfo.mTtl, compressInfo));
+            IncResourceRecordCount(responseHeader, additional);
+        }
+
+        if (HasQuestion(aQuery.GetResponseHeader(), aQuery.GetResponseMessage(), aInstanceInfo.mHostName,
+                        ResourceRecord::kTypeAaaa) == !additional)
+        {
+            for (uint8_t i = 0; i < aInstanceInfo.mAddressNum; i++)
+            {
+                const Ip6::Address &address = static_cast<const Ip6::Address &>(aInstanceInfo.mAddresses[i]);
+
+                OT_ASSERT(!address.IsUnspecified() && !address.IsLinkLocal() && !address.IsMulticast() &&
```

Comment:
Are we assuming that the querier will ensure that such addresses are never provided?

Reply:
Yes we are. We are trying to be strict in DNS-SD server when input from Discovery Proxy implementations so as to catch bugs earlier.

Reply:
Should we call out these constraints in the API docs?